instrumentation/servlet/request-2.3/src/main/java/io/opentelemetry/auto/instrumentation/servlet/v2_3/Servlet2Advice.java

```diff
@@ -90,40 +55,17 @@ public static void stopSpan(
       @Advice.Argument(1) final ServletResponse response,
       @Advice.Enter final SpanWithScope spanWithScope,
       @Advice.Thrown final Throwable throwable) {
-    // Set user.principal regardless of who created this span.
-    final Object spanAttr = request.getAttribute(SPAN_ATTRIBUTE);
-    if (spanAttr instanceof Span && request instanceof HttpServletRequest) {
-      final Principal principal = ((HttpServletRequest) request).getUserPrincipal();
-      if (principal != null) {
-        ((Span) spanAttr).setAttribute(MoreTags.USER_NAME, principal.getName());
-      }
-    }
+    if (request instanceof HttpServletRequest && response instanceof HttpServletResponse) {
+      TRACER.setPrincipal((HttpServletRequest) request);
 
-    if (spanWithScope == null) {
-      return;
-    }
-
-    final Span span = spanWithScope.getSpan();
-
-    if (response instanceof HttpServletResponse) {
-      DECORATE.onResponse(
-          span, InstrumentationContext.get(ServletResponse.class, Integer.class).get(response));
-    } else {
-      DECORATE.onResponse(span, null);
-    }
+      Integer responseStatus =
+          InstrumentationContext.get(ServletResponse.class, Integer.class).get(response);
 
-    if (throwable != null) {
-      if (response instanceof HttpServletResponse
-          && InstrumentationContext.get(ServletResponse.class, Integer.class).get(response)
-              == HttpServletResponse.SC_OK) {
-        // exception was thrown but status code wasn't set
-        span.setAttribute(Tags.HTTP_STATUS, 500);
-        span.setStatus(Status.UNKNOWN);
+      if (throwable == null) {
+        TRACER.end(spanWithScope, responseStatus);
+      } else {
+        TRACER.endExceptionally(spanWithScope, throwable, responseStatus);
```

Comment:
something like this is what i was trying to say
```suggestion
        if (!response.isCommitted()) {
          responseStatus = 500;
        }
        TRACER.endExceptionally(spanWithScope, throwable, responseStatus);
```

Reply:
Do you suggest to add this condition check to _all_ servlet instrumentations who calls this method? This means `Servlet2Advice`, `Servlet3Advice` and `TagSettingAsyncListener`. Avoiding such duplication was the original goal of this PR :)

In addition, should we check documentation for all other http servers/frameworks, if they require the same logic?

Now thinking about this... Your citation from servlet spec says about changing status code of the servlet response. We don't do this here and did not do that before. So this current logic will report response status code in Span which is different from the one that client got in response. This was my original concern and why I think this is wrong.

Reply:
> Do you suggest to add this condition check to _all_ servlet instrumentations who calls this method? This means `Servlet2Advice`, `Servlet3Advice` and `TagSettingAsyncListener`. Avoiding such duplication was the original goal of this PR :)

Oh yes 😄. How about a subclass of `HttpServerTracer` that handles servlet-specific stuff?

> Now thinking about this... Your citation from servlet spec says about changing status code of the servlet response. We don't do this here and did not do that before. So this current logic will report response status code in Span which is different from the one that client got in response. This was my original concern and why I think this is wrong.

Adding this logic (checking `isCommitted`) should make the response code in the Span align with the response code received by the client, no?

Reply:
> Adding this logic (checking `isCommitted`) should make the response code in the Span align with the response code received by the client, no?

I think not. We read status code from the response. Then if that status code is not 200 and exception was thrown, then we set `http.status_code` attribute value to 500. But we don't change response status code in `HttpResponse` object. So client will still get 200. Or am I wrong?

Reply:
Ah, no, this part of the servlet spec is defining how _servlet containers_ behave:

> If the exception isn't caught and handled, and the response hasn’t been committed, the status code MUST be set to 500

So if an exception is thrown, and the response has not been committed, the _servlet container_ will change the response code to 500 and the client will get 500.

Reply:
> How about a subclass of HttpServerTracer that handles servlet-specific stuff?

I already have one in this PR, `ServletHttpServerTracer`. But it still means that I have to bring back `ResponseWithStatus`